pkg/event/registry_test.go

```diff
@@ -0,0 +1,108 @@
+package event_test
+
+import (
+	"errors"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	k8sV1 "k8s.io/api/core/v1"
+	v1meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	submV1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/event"
+	"github.com/submariner-io/submariner/pkg/event/logger"
+)
+
+const npOVNKubernetes = "OVNKubernetes"
+const npGenericKubeproxyIptables = "GenericKubeproxyIptables"
+
+var _ = Describe("Event Registry", func() {
+	When("handlers with different network plugin are added to the registry", func() {
+		var (
+			ovn, generic *TestHandler
+			registry     event.Registry
+		)
+
+		BeforeEach(func() {
+			registry = event.NewRegistry("test-registry", npGenericKubeproxyIptables)
+			ovn = NewTestHandler("handler-ovn", npOVNKubernetes)
+			generic = NewTestHandler("handler-kubeproxy", npGenericKubeproxyIptables)
+			err := registry.AddHandlers(logger.NewHandler(), ovn, generic)
+			Expect(err).NotTo(HaveOccurred())
+			cleanupTestEvents()
+		})
+
+		It("should only initialize the handlers whose network plugin matches that of the registry", func() {
+			Expect(ovn.Initialized).To(BeFalse())
+			Expect(generic.Initialized).To(BeTrue())
+		})
+
+		It("should only invoke handlers whose network plugin matches that of the registry", func() {
+			_ = registry.TransitionToGateway()
+			Expect(ovn.Events).To(HaveLen(0))
+			Expect(generic.Events).To(HaveLen(1))
+		})
+	})
+
+	When("handlers with the same network plugin are added to the registry", func() {
+		var (
+			handler1, handler2 *TestHandler
+			registry           event.Registry
+		)
+
+		BeforeEach(func() {
+			registry = event.NewRegistry("test-registry", npGenericKubeproxyIptables)
+			handler1 = NewTestHandler("handler1", npGenericKubeproxyIptables)
+			handler2 = NewTestHandler("handler2", npGenericKubeproxyIptables)
+			err := registry.AddHandlers(logger.NewHandler(), handler1, handler2)
+			Expect(err).NotTo(HaveOccurred())
+			cleanupTestEvents()
+		})
+
+		It("should initialize all the handlers", func() {
+			Expect(handler1.Initialized).To(BeTrue())
+			Expect(handler2.Initialized).To(BeTrue())
+		})
+
+		It("Should be called in registration order", func() {
+			_ = registry.TransitionToGateway()
+			Expect(allTestEvents).Should(Receive(Equal(TestEvent{Name: evTransitionToGateway, Handler: handler1.Name})))
+			Expect(allTestEvents).Should(Receive(Equal(TestEvent{Name: evTransitionToGateway, Handler: handler2.Name})))
```

Comment:
To be complete we should test this for all events. I suggest removing this spec and adding these checks to the spec below.

Reply:
We can do that, yup